src/main/kotlin/no/fdk/fdk_reasoning_service/kafka/KafkaHarvestedEventCircuitBreaker.kt

```diff
@@ -0,0 +1,72 @@
+import no.fdk.concept.ConceptEvent
+import no.fdk.concept.ConceptEventType
+import no.fdk.dataservice.DataServiceEvent
+import no.fdk.dataservice.DataServiceEventType
+import no.fdk.dataset.DatasetEvent
+import no.fdk.dataset.DatasetEventType
+import no.fdk.event.EventEvent
+import no.fdk.event.EventEventType
+import no.fdk.fdk_reasoning_service.model.CatalogType
+import no.fdk.informationmodel.InformationModelEvent
+import no.fdk.informationmodel.InformationModelEventType
+import no.fdk.service.ServiceEvent
+import no.fdk.service.ServiceEventType
+import org.apache.avro.specific.SpecificRecord
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+@Component
+class KafkaHarvestedEventCircuitBreaker {
+    @CircuitBreaker(name = "reasoning")
+    fun process(record: ConsumerRecord<String, SpecificRecord>) {
+        LOGGER.debug("Received message - offset: ${record.offset()}")
+        val event = record.value()
+
+        try {
+            val (fdkId, graph, timestamp, resourceType) = getKafkaEventData(event)
+            LOGGER.debug("Reason {} - id: {}", resourceType, fdkId)
+            reasonAndProduceEvent(fdkId, graph, timestamp, resourceType)
+        } catch (e: Exception) {
+            LOGGER.error("Error occurred during reasoning: " + e.message)
+            throw e
+        }
+    }
+
+    private fun getKafkaEventData(event: SpecificRecord): EventData {
+        event.let {
+            if (it is DatasetEvent && it.type == DatasetEventType.DATASET_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.DATASETS)
+            } else if (it is ConceptEvent && it.type == ConceptEventType.CONCEPT_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.CONCEPTS)
+            } else if (it is DataServiceEvent && it.type == DataServiceEventType.DATA_SERVICE_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.DATASERVICES)
+            } else if (it is InformationModelEvent && it.type == InformationModelEventType.INFORMATION_MODEL_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.INFORMATIONMODELS)
+            } else if (it is ServiceEvent && it.type == ServiceEventType.SERVICE_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.PUBLICSERVICES)
+            } else if (it is EventEvent && it.type == EventEventType.EVENT_HARVESTED) {
+                return EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.EVENTS)
+            } else {
+                LOGGER.warn("Received unknown kafka event type")
+                throw Exception("Error when reading kafka event")
+            }
```

Comment:
```suggestion
    private fun getKafkaEventData(event: SpecificRecord): EventData? =
        when {
            event is DatasetEvent && event.type == DatasetEventType.DATASET_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.DATASETS)
            event is ConceptEvent && event.type == ConceptEventType.CONCEPT_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.CONCEPTS)
            event is DataServiceEvent && event.type == DataServiceEventType.DATA_SERVICE_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.DATASERVICES)
            event is InformationModelEvent && event.type == InformationModelEventType.INFORMATION_MODEL_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.INFORMATIONMODELS)
            event is ServiceEvent && event.type == ServiceEventType.SERVICE_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.PUBLICSERVICES)
            event is EventEvent && event.type == EventEventType.EVENT_HARVESTED) ->
               EventData(it.fdkId.toString(), it.graph.toString(), it.timestamp, CatalogType.EVENTS)
            else -> LOGGER.debug("Received unsupported kafka event")
```

Reply:
er ikke det enklere å lese?

Reply:
Jo, bra forslag! Prøvde meg først på `when`, men fikk det ikke til å funke. Men dette ser bra ut.

Reply:
Ikke nødvendig med when, men jeg synes det er lettere å lese. Men er to ting her du må fikse:
- bruken av let-funksjonen gir ingen positiv verdi og gjør det bare vanskeligere å lese
- det er for mange `return`, best hvis du klarer å holde deg til en